### ISSUE TITLE
auth_basic doesn't allow unauthorized users

### ISSUE THREAD

**Issue by Bystroushaak** [bottlepy/bottle#706]
I am not really sure that this is a bug, but [@auth_basic()](https://github.com/bottlepy/bottle/blob/master/bottle.py#L2573) decorator doesn't allow unauthorized users even in case, that authorization function say that it is okay.

The "problem" is in line with condition, which checks existence of username before it checks the autorization function ``check()``.

```python
if user is None or not check(user, password):
```

I believe, that it should only check the authorization function, because otherwise you can't have routes that are accessible to all, but they still have ``@auth_basic()`` decorator. This may be quite annoying while debugging your authorization function and also there are some use-cases when you do something more advanced in your authorization function, like different access levels for each path and so on.

Tl dr; I think that user's authorization function should choose whether to authorize or reject user, not the presence of ``@auth_basic()`` decorator itself.
